usecase/call_rpc.go

```diff
@@ -96,17 +98,54 @@ func (m *dependencyManager) CallRPC(ctx context.Context, w io.Writer, rpcName st
 		return flushDone()
 	}
 
-	md := metadata.New(nil)
-	for k, v := range m.ListHeaders() {
-		md.Append(k, v...)
+	parseDuration := func(duration string) (time.Duration, error) {
+		duration = strings.Replace(duration, "n", "ns", 1)
+		duration = strings.Replace(duration, "u", "us", 1)
+		duration = strings.Replace(duration, "m", "ms", 1)
+		duration = strings.Replace(duration, "S", "s", 1)
+		duration = strings.Replace(duration, "M", "m", 1)
+		duration = strings.Replace(duration, "H", "h", 1)
+		timeout, err := time.ParseDuration(duration)
+		if err != nil {
+			return 0, errors.Wrapf(err, "Malformed grpc-timeout header")
+		}
+		return timeout, err
+	}
+
+	enhanceContext := func(ctx context.Context) (context.Context, error) {
+		md := metadata.New(nil)
+		for k, v := range m.ListHeaders() {
+			md.Append(k, v...)
+		}
+
+		ctx = metadata.NewOutgoingContext(ctx, md)
+		values := md.Get("grpc-timeout")
+		if len(values) == 0 {
+			return ctx, nil
+		}
+
+		timeout, err := parseDuration(values[len(values)-1])
+		if err != nil {
+			return nil, err
+		}
+
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+
+		go func() {
+			t := time.NewTicker(timeout)
+			<-t.C
+			cancel()
+		}()
```

Comment:
Could you return `cancel` function instead of calling `cancel()` by using a goroutine?
I think it's a similar signature to WithTimeout, so it's more intuitive.